numpy/lib/twodim_base.py

```diff
@@ -32,6 +32,75 @@ def _min_int(low, high):
     return int64
 
 
+def flip(m, axis):
+    """
+    Reverse the order of elements in an array along the given axis.
+
+    The shape of the array is preserved, but the elements are reordered.
+
+    Parameters
+    ----------
+    m : array_like
+        Input array.
+    axis: integer
+        Axis in array, which entries are reversed.
+        
+
+    Returns
+    -------
+    out : array_like
+        A view of `m` with the entries of axis reversed.  Since a view is
+        returned, this operation is :math:`\\mathcal O(1)`.
+
+    See Also
+    --------
+    flipud : Flip an array vertically (axis=0).
+    fliplr : Flip an array horizontally (axis=1).
+
+    Notes
+    -----
+    flip(m, 0) is equivalent to flipud(m).
+    flip(m, 1) is equivalent to fliplr(m).
+    flip(m, n) corresponds to ``A[...,::-1,...]`` with ``::-1`` at position n.
+
+    Examples
+    --------
+    >>> A = np.arange(8).reshape((2,2,2))
+    >>> A
+    array([[[0, 1],
+            [2, 3]],
+
+           [[4, 5],
+            [6, 7]]])
+
+    >>> flip(A, 0)
+    array([[[4, 5],
+            [6, 7]],
+    
+           [[0, 1],
+            [2, 3]]])
+
+    >>> flip(A, 1)
+    array([[[2, 3],
+            [0, 1]],
+    
+           [[6, 7],
+            [4, 5]]])
+
+    >>> A = np.random.randn(3,4,5)
+    >>> np.all(flip(A,2)==A[:,:,::-1,...])
+    True
+    """
+    if not hasattr(m, 'ndim'):
+        m = asarray(m)
+    if m.ndim <= axis:
```

Comment:
No, ValueError is definitely the right error to raise for an invalid axis.
On Fri, Feb 26, 2016 at 9:30 PM Marten van Kerkwijk <
notifications@github.com> wrote:

> In numpy/lib/twodim_base.py
> <https://github.com/numpy/numpy/pull/7346#discussion_r54326237>:
>
> > +            [2, 3]]])
> > +
> > +    >>> flip(A, 1)
> > +    array([[[2, 3],
> > +            [0, 1]],
> > +
> > +           [[6, 7],
> > +            [4, 5]]])
> > +
> > +    >>> A = np.random.randn(3,4,5)
> > +    >>> np.all(flip(A,2)==A[:,:,::-1,...])
> > +    True
> > +    """
> > +    if not hasattr(m, 'ndim'):
> > +        m = asarray(m)
> > +    if m.ndim <= axis:
>
> This doesn't cover negative values of axis. I would remove this
> altogether and replace it by making the assignment part of a try/except, as
> follows:
>
>     try:
>         indexer[axis] = slice(None, None, -1)
>     except IndexError:
>         raise ValueError(...)
>
> Question: Should this be an IndexError rather than a ValueError?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7346/files#r54326237>.
>